File: RAII/ruby/safe_resource.rb
class SafeResource
  def self.acquire
    resource = self.new
    yield resource
  ensure
    resource.release
  end

  def initialize
    puts "Acquiring resource"
    @resource = Object.new
  end

  def release
    puts "Releasing resource"
    @resource = nil
  end
end

File: RAII/c++/SafeResource.cpp
#include "SafeResource.h"
#include "support.h"

SafeResource::SafeResource() :
    resource(acquireResource())
{
}

SafeResource::~SafeResource()
{
    releaseResource(resource);
}

Resource* SafeResource::get()
{
    return resource;
}
